File: sdks/dart/lib/src/edge.dart
import 'package:dio/dio.dart';

class Edge {
  final Dio _dio;

  Edge(this._dio);

  Future<Map<String, dynamic>> create({
    required String name,
    required String code,
    String? route,
    int? timeout,
  }) async {
    final res = await _dio.post('/v1/edge/functions', data: {
      'name': name,
      'code': code,
      if (route != null) 'route': route,
      if (timeout != null) 'timeout': timeout,
    });
    return res.data;
  }

  Future<Map<String, dynamic>> list() async {
    final res = await _dio.get('/v1/edge/functions');
    return res.data;
  }

  Future<Map<String, dynamic>> get(String functionId) async {
    final res = await _dio.get('/v1/edge/functions/$functionId');
    return res.data;
  }

  Future<Map<String, dynamic>> update(
    String functionId, {
    String? name,
    String? code,
    String? route,
    int? timeout,
  }) async {
    final res = await _dio.put('/v1/edge/functions/$functionId', data: {
      if (name != null) 'name': name,
      if (code != null) 'code': code,
      if (route != null) 'route': route,
      if (timeout != null) 'timeout': timeout,
    });
    return res.data;
  }

  Future<void> delete(String functionId) async {
    await _dio.delete('/v1/edge/functions/$functionId');
  }

  Future<Map<String, dynamic>> invoke(
    String functionId, {
    Map<String, dynamic>? data,
  }) async {
    final res = await _dio.post('/v1/edge/functions/$functionId/invoke',
        data: data ?? {});
    return res.data;
  }

  Future<Map<String, dynamic>> listExecutions(String functionId) async {
    final res = await _dio.get('/v1/edge/functions/$functionId/executions');
    return res.data;
  }
}

File: sdks/node/src/edge.ts
import type { ApplAdServer } from './client';

/**
 * Edge functions.
 *
 * EXPERIMENTAL / NOT IMPLEMENTED on this build: the backend Edge service is
 * not yet wired up, so `invoke`, `create` and the other methods here currently
 * return HTTP 501 (Not Implemented). The API surface is kept stable for forward
 * compatibility; do not rely on it in production yet.
 */
export class Edge {
  constructor(private client: ApplAdServer) {}

  create(name: string, code: string, opts?: { route?: string; timeout?: number }) {
    return this.client.call('POST', '/edge/functions', { name, code, ...opts });
  }

  list() {
    return this.client.call('GET', '/edge/functions');
  }

  get(functionId: string) {
    return this.client.call('GET', `/edge/functions/${functionId}`);
  }

  update(functionId: string, opts: { name?: string; code?: string; route?: string; timeout?: number }) {
    return this.client.call('PUT', `/edge/functions/${functionId}`, opts);
  }

  delete(functionId: string) {
    return this.client.call('DELETE', `/edge/functions/${functionId}`);
  }

  invoke(functionId: string, data?: Record<string, unknown>) {
    return this.client.call('POST', `/edge/functions/${functionId}/invoke`, data ?? {});
  }

  listExecutions(functionId: string) {
    return this.client.call('GET', `/edge/functions/${functionId}/executions`);
  }
}
